Update median doc to note return value of even-sized input

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44562 Update median doc to note return value of even-sized input**

Add a note that torch.median returns the smaller of the two middle elements for even-sized input and refer user to torch.quantile for the mean of the middle values.

fixes https://github.com/pytorch/pytorch/issues/39520

Differential Revision: [D23657208](https://our.internmc.facebook.com/intern/diff/D23657208)

heitorschueroff committed:
Update median doc to note return value of even-sized input

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4530,6 +4530,11 @@ def merge_dicts(*dicts):
 Returns the median value of all elements in the :attr:`input` tensor.
 
+.. note::
+    For even-sized inputs, the smaller of the two middle elements in the sorted
```
I'm not in love with saying "smaller." For example:

```
t = torch.tensor((-3, -5))
torch.median(t)
: tensor(-5)
```

Is -5 "smaller" than -3? I feel like that's debatable at best. What about using the term "lower"? That would be consistent with std::numeric_limits, for example.

"Middle elements" is also a little vague. I would appeal to the more formal definition of the median https://en.wikipedia.org/wiki/Median, say it may not be unique, and that when the median is not unique we pick the lower of the two medians in the input.

mruberry replied:
"even-sized inputs" is a problem. In the first definition of median the tensor is flatted, so it's inputs with an even number of elements. In the second definition of median the median is computed along its a dimension, so it's whether that dimension has an even size.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4530,6 +4530,11 @@ def merge_dicts(*dicts):
 
+.. note::
+    For even-sized inputs, the smaller of the two middle elements in the sorted
+    order is returned. If the mean of the two middle elements is required, use
```
If the above suggestion is adopted, this would be something like, "To compute the mean of both medians in the input, use...".

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4530,6 +4530,11 @@ def merge_dicts(*dicts):
 
 Returns the median value of all elements in the :attr:`input` tensor.
 
+.. note::
```
This note is a little wonky because median, like other reductions, is documented twice. Worse, it's really two different functions: median and median_with_indices, which is definitely an API bug that we shouldn't try to address in this PR. Unfortunately, it means you probably need to duplicate this note for both versions of the documentation.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4530,6 +4530,11 @@ def merge_dicts(*dicts):
+    :func:`torch.quantile` with ``q=0.5`` instead.
+
 .. warning::
     This function produces deterministic (sub)gradients unlike ``median(dim=0)``
```
Aside: it's very surprising to me that this version of median produces deterministic gradients but the other version, which actually computes indices, doesn't. We should probably fix "median_with_indices" to consistently return the FIRST valid median, just like we fixed argmin and argmax.

albanD replied:
I think, as you mention, that the other function returns non-deterministic indices and thus non-deterministic subgradients.
This one evenly distribute the gradient to all the inputs with the value used. So it always is deterministic.

heitorschueroff replied:
The reason it returns non-deterministic indices is because it uses quickselect to partition which is not stable. One way to solve it is to use 3-way quickselect that partitions input into <, = and >. Then choose the smallest index from the = segment.

heitorschueroff committed:
Update on "Update median doc to note return value of even-sized input"


Add a note that torch.median returns the smaller of the two middle elements for even-sized input and refer user to torch.quantile for the mean of the middle values.

fixes https://github.com/pytorch/pytorch/issues/39520

Differential Revision: [D23657208](https://our.internmc.facebook.com/intern/diff/D23657208)

[ghstack-poisoned]
